File: ricemill/ricemill/custom/js/job_card.js
frappe.ui.form.on("Job Card", {
    "refresh": function (frm) {
        if (!frm.doc.employee.length) {
            frm.set_value('time_logs', [])
        }
        frm.add_custom_button(__(""), () => {
            let data = [];
            var item = frm.doc.production_item
            const fields = [
                {
                    label: "Items",
                    fieldtype: "Table",
                    fieldname: "items",
                    cannot_add_rows: true,
                    in_place_edit: true,
                    data: data,
                    get_data: () => {
                        return data;
                    },
                    fields: [
                        {
                            fieldtype: "Data",
                            fieldname: "docname",
                            hidden: true
                        },
                        {
                            fieldtype: "Read Only",
                            fieldname: "item_code",
                            label: __("Item Code"),
                            in_list_view: true,
                            options: "Item",
                        },
                        {
                            fieldtype: "Read Only",
                            fieldname: "item_name",
                            label: __("Item Name"),
                            in_list_view: true
                        },
                        {
                            fieldtype: "Float",
                            fieldname: "qty",
                            label: __("Accepted Quantity"),
                            in_list_view: true,
                            read_only: true
                        },
                        {
                            fieldtype: "Float",
                            fieldname: "sample_size",
                            label: __("Sample Size"),
                            reqd: true,
                            in_list_view: true
                        },
                        {
                            fieldtype: "Data",
                            fieldname: "description",
                            label: __("Description"),
                            hidden: true
                        },
                        {
                            fieldtype: "Data",
                            fieldname: "serial_no",
                            label: __("Serial No"),
                            hidden: true
                        },
                        {
                            fieldtype: "Data",
                            fieldname: "batch_no",
                            label: __("Batch No"),
                            hidden: true
                        }
                    ]
                }
            ];

            const dialog = new frappe.ui.Dialog({
                title: __("Select Items for Quality Inspection"),
                fields: fields,
                primary_action: function () {
                    const data = dialog.get_values();
                    frappe.call({
                        method: "erpnext.controllers.stock_controller.make_quality_inspections",
                        args: {
                            doctype: frm.doc.doctype,
                            docname: frm.doc.name,
                            items: data.items
                        },
                        freeze: true,
                        callback: function (r) {
                            if (r.message.length > 0) {
                                if (r.message.length === 1) {
                                    frappe.set_route("Form", "Quality Inspection", r.message[0]);
                                } else {
                                    frappe.route_options = {
                                        "reference_type": frm.doc.doctype,
                                        "reference_name": frm.doc.name
                                    };
                                    frappe.set_route("List", "Quality Inspection");
                                }
                            }
                            dialog.hide();
                        }
                    });
                },
                primary_action_label: __("Create")
            });
            let dialog_items = dialog.fields_dict.items;
            frappe.db.get_list("BOM Item", { filters: { 'parent': frm.doc.bom_no }, fields: ["item_code", "item_name"] }).then((bom_item) => {
                bom_item.forEach((item) => {
                    dialog_items.df.data.push({
                        "docname": frm.doc.name,
                        "item_code": item.item_code,
                        "item_name": item.item_name,
                    });
                    dialog_items.grid.refresh();
                    data = dialog.fields_dict.items.df.data;
                    if (!data.length) {
                        frappe.msgprint(__("All items in this document already have a linked Quality Inspection."));
                    } else {
                        dialog.show();
                    }
                })
            })

        }, __("Create"));
        frappe.db.get_list("Operation", { fields: ['hulling'], filters: { 'name': cur_frm.doc.operation } }).then((value) => {
            if (value[0].hulling === 1) {
                frm.add_custom_button(__("Stock Entry"), () => {
                    frappe.call({
                        method: "ricemill.ricemill.custom.py.job_card.repack_dialog",
                        async: false,
                        freeze: true,
                        freeze_message: "Please wait...",
                        args: {
                            item: frm.doc.production_item,
                            operation: frm.doc.operation,
                            qty: frm.doc.total_completed_qty
                        },
                        callback: function (r) {
                            if (r.message[0]) {
                                let count = 0
                                var func = function () {
                                    var used_qty = 0;
                                    for (let i = 1; i <= r.message[1]; i++) {
                                        used_qty += d.fields_dict['qty' + i].value * d.fields_dict['conv' + i].value
                                    }
                                    d.fields_dict['remaining_qty'].value = d.fields_dict.total_qty.value - used_qty
                                    d.fields_dict['remaining_qty'].refresh()

                                }
                                r.message[0].forEach((data) => {
                                    if (data['fieldname'].indexOf('qty') >= 0 && data['fieldname'] != 'total_qty' && data['fieldname'] != 'remaining_qty') {
                                        r.message[0][count]['onchange'] = func
                                    }
                                    count += 1
                                })
                                var d = new frappe.ui.Dialog({
                                    title: "Repack details for hulling operation",
                                    fields: r.message[0],
                                    freeze: true,
                                    static: true,
                                    primary_action: function (data) {
                                        if (d.fields_dict.remaining_qty.value === 0 || d.fields_dict.remaining_qty.value === null) {
                                            frappe.call({
                                                method: "ricemill.ricemill.custom.py.job_card.make_repack",
                                                args: {
                                                    items: data,
                                                    length: r.message[1],
                                                    work_order_id: cur_frm.doc.work_order,
                                                    purpose: cur_frm.doc.ts_stock_entry_type,
                                                    qty: cur_frm.doc.total_completed_qty,
                                                    sw: cur_frm.doc.ts_source_warehouse,
                                                    tw: cur_frm.doc.ts_target_warehouse
                                                },
                                                callback(r) {
                                                    frappe.model.sync(r.message);
                                                    frappe.set_route('Form', r.message.doctype, r.message.name);
                                                }
                                            })
                                            d.hide();
                                        }
                                        else {
                                            frappe.throw("Remaining qty must be equal to zero")
                                        }
                                    }
                                })
                                d.show()
                            }
                        }
                    })
                }, __("Create"));
            }
        });
    }
});

File: ricemill/ricemill/custom/py/job_card.py
from curses import KEY_MARK
import frappe
from frappe.utils import get_link_to_form
import json
from frappe.utils.data import flt
@frappe.whitelist()
def check_quality_inspection(doc):
    doc=json.loads(doc)
    operation=frappe.get_all("Operation",{'name':doc['operation']},["grading","soaking","boiling","dryer","hulling"])
    opr=""
    for i in operation[0]:
        if operation[0][i]==1:
            opr=i
            break
    if opr=="":frappe.throw("Select a Valid Operation  "+ get_link_to_form("Operation",doc['operation']))
    inspection_req=frappe.get_value("Item",doc['production_item'],opr)
    if inspection_req:
        quality_inspection=frappe.get_value("Quality Inspection",{'reference_name':doc['name']},"name")
        if not quality_inspection:
            frappe.throw ("Quality Inspection is Missing")
def before_submit(doc,action):
    jc_rej=sum([flt(i) for i in frappe.get_all("Quality Inspection",{'reference_name':doc.name,'status':'Rejected'},pluck='accepted_rejected_qty')])
    if jc_rej:
        frappe.msgprint(f'The Rejected item Quantity {jc_rej}')

@frappe.whitelist()
def repack_dialog(item,operation,qty):
    opr  = frappe.db.get_value("Operation", operation, 'hulling')
    if(not opr):return
    else:
        uom = frappe.get_all("UOM Conversion Detail", {'parent':item},['uom','conversion_factor'])
        fields=[]
        stock_uom = frappe.get_value("Item", item, 'stock_uom')
        for i in range(len(uom)):
            fields.append({'fieldname':f'uom{i+1}','label':uom[i]['uom'],'fieldtype':'Link','default':uom[i]['uom'],'read_only':1})
            fields.append({'fieldtype':'Column Break','fieldname':f'column{i+1}'})
            fields.append({'fieldtype':'Float','fieldname':f'conv{i+1}','default':uom[i]['conversion_factor'],'read_only':1, 'hidden':1})
            fields.append({'fieldtype':'Column Break','fieldname':f'column1{i+1}'})
            fields.append({'fieldtype':'Float','fieldname':f'qty{i+1}','default':0,'description':f"Enter QTY in {stock_uom}"})
            fields.append({'fieldtype':'Section Break','fieldname':f'section{i+1}'})
        fields+=[
            {'fieldname':f'default_uom','label':"Stock UOM",'fieldtype':'Link','default':stock_uom,'read_only':1},
            {'fieldtype':'Column Break','fieldname':'column_1'},
            {'fieldtype':'Float','label':"Total QTY",'fieldname':'total_qty','default':qty,'read_only':1},
            {'fieldtype':'Column Break','fieldname':f'column_2'},
            {'fieldtype':'Float','label':"Remaining QTY",'fieldname':'remaining_qty','default':qty},
        ]
        return fields, len(uom)

@frappe.whitelist()
def make_repack(items, length, work_order_id, purpose, qty=0, sw=None, tw = None):
    items = json.loads(items)
    uom,conv,qnty = separate_items_data(items, length)
    work_order = frappe.get_doc("Work Order", work_order_id)
    se_item = ''
    se_item = frappe.get_doc("Item",work_order.production_item)
    parent_wo = frappe.get_value("Work Order", work_order_id, 'ts_parent_work_order')
    se_name = frappe.get_value("Stock Entry", {'ts_work_order':parent_wo},'name')
    batch_no = frappe.get_value("Batch",{'reference_doctype':'Stock Entry','reference_name':se_name},'name')
    scrap_item = ''
    stock_entry = frappe.new_doc("Stock Entry")
    stock_entry.ts_work_order = work_order_id
    stock_entry.purpose = purpose
    stock_entry.company = work_order.company
    items ={ 
        "item_code":se_item.item_code,
        "item_group": se_item.item_group, 
        "item_name":se_item.item_name, 
        "qty": qty , 
        "uom":se_item.stock_uom, 
        "conversion_factor": 1,
        "batch_no": batch_no,
        "s_warehouse" : sw
         }
    stock_entry.append("items", items)

    ## Add Repacked item in different uoms
    for i in range(len(qnty)):
        items ={ 
        "item_code":se_item.item_code,
        "item_group": se_item.item_group, 
        "item_name":se_item.item_name, 
        "qty": qnty[i] , 
        "uom":uom[i], 
        "conversion_factor": conv[i],
        "t_warehouse": tw,
        "is_finished_item": 1
         }
        stock_entry.append("items", items)
    # stock_entry.set_serial_no_batch_for_finished_good()
    
    from erpnext.manufacturing.doctype.bom.bom import add_additional_cost
    add_additional_cost(stock_entry, work_order)
    
    job_card = frappe.get_all("Job Card", filters={'work_order':work_order_id},pluck = 'name')[0]
    jc_doc = frappe.get_doc("Job Card", job_card)
    scrap_item = jc_doc.scrap_items
    for i in scrap_item:
        items ={ 
        "item_code":i.item_code,
        "item_group": frappe.get_value("Item", i.item_code, 'item_group'), 
        "item_name":i.item_name, 
        "qty": i.stock_qty , 
        "uom":i.stock_uom, 
        "conversion_factor": 1,
        "is_scrap_item": 1,
        "t_warehouse" : tw,
        }
        if(purpose != 'Manufacture'):items['batch_no'] = batch_no
        stock_entry.append("items", items)
    stock_entry.project = work_order.project
    stock_entry.set_stock_entry_type()
    stock_entry.insert(ignore_mandatory=True, ignore_permissions=True)
    stock_entry.submit()
    jc_doc.submit()
    from ricemill.ricemill.custom.py.workorder import change_status
    change_status(work_order_id)
    frappe.db.commit()
    return stock_entry.as_dict()


def separate_items_data(items, length):
    uom, conv, qty= [], [], []
    for i in range(1, int(length)+1):
        if(items[f"qty{i}"] > 0):
            uom.append(items[f"uom{i}"])
            conv.append(items[f"conv{i}"])
            qty.append(items[f"qty{i}"])
    return uom,conv,qty